api/webapp.go

```diff
@@ -59,15 +61,57 @@ func (s Server) FrontendHandler(prefix string) http.Handler {
 	// TODO -
 	// * there's no error handling,
 	// * and really the data could be owned by Server and initialized by it,
-	//   as there's nothing that necessitates updating the webapp within FrontendHandler.
-	if err := s.resolveWebappPath(); err != nil {
-		log.Errorf("error resolving webapp path: %s", err)
-	}
+	//  as there's nothing that necessitates updating the webapp within FrontendHandler.
+	updating := true
+	updates := make(chan bool)
+	go func() {
+		if err := s.resolveWebappPath(); err != nil {
+			log.Errorf("error resolving webapp path: %s", err)
+		}
+		path := s.Config().Webapp.EntrypointHash
+
+		// TODO (ramfox): frontend handler should accept a context
+		log.Info("fetching webapp off the distributed web...")
+		fetcher, ok := s.Repo().Store().(cafs.Fetcher)
+		if !ok {
+			log.Error("this store cannot fetch from remote sources")
+			updates <- false
+			return
+		}
+		_, err := fetcher.Fetch(cafs.SourceAny, path+"/main.js")
+		if err != nil {
+			log.Errorf("error fetching file: %s", err.Error())
+			updates <- false
+		}
+		log.Info("pinning webapp to your repository...")
+		pinner, ok := s.Repo().Store().(cafs.Pinner)
+		if !ok {
+			log.Error("unable to pin webapp to your repository")
+			updates <- false
+			return
+		}
+		if err := pinner.Pin(path, true); err != nil {
+			log.Errorf("error pinning webapp: %s", err.Error())
+			updates <- false
+			return
+		}
+		log.Info("done pinning webapp!")
+		updates <- false
+	}()
+
+	go func() {
+		updating = <-updates
+	}()
+
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if updating == true {
+			// if the app is still pinning
+			// return a temporary script that tells the user what's going on
+			w.Write([]byte(loadingScript))
+			return
+		}
 		path := fmt.Sprintf("%s%s", s.Config().Webapp.EntrypointHash, strings.TrimPrefix(r.URL.Path, prefix))
-		log.Info("fetching webapp off the distributed web...")
 		s.fetchCAFSPath(path, w, r)
```

Comment:
With the errors channel change from above in place, we'll now need to check if the value of `fetchErr` is non-ni here (after confirming `updating` is false, before `fetchCAFSPath`), and print an error for the user if so.